checker/src/fixed_point_visitor.rs

```diff
@@ -146,38 +149,46 @@ impl<'fixed, 'analysis, 'compilation, 'tcx, E>
                 self.bv.function_name
             );
         }
+        last_block
     }
 
     /// Visits a loop body. Return true if the out_state computed by this visit is not a subset
     /// of the out_state computed previously. When it is a subset, a fixed point has been reached.
     /// A loop body is all of the blocks that are dominated by the loop anchor.
     #[logfn_inputs(TRACE)]
-    fn visit_loop_body(&mut self, loop_anchor: mir::BasicBlock, iteration_count: usize) -> bool {
+    fn visit_loop_body(
+        &mut self,
+        loop_anchor: mir::BasicBlock,
+        iteration_count: usize,
+    ) -> (bool, mir::BasicBlock) {
         let mut changed = false;
+        let mut last_block = loop_anchor;
         let blocks = self.block_indices.clone();
+        let old_state = self.out_state.clone();
         for bb in blocks {
             if !self.already_visited.contains(&bb)
                 && self.dominators.is_dominated_by(bb, loop_anchor)
             {
+                last_block = bb;
                 // Visit the next block, or the entire nested loop anchored by it.
                 if bb == loop_anchor {
                     self.visit_basic_block(bb, iteration_count); // join or widen
                 } else if self.loop_anchors.contains(&bb) {
-                    self.compute_fixed_point(bb);
+                    last_block = self.compute_fixed_point(bb);
                 } else {
                     self.visit_basic_block(bb, 0); // conditional expressions
                 }
 
                 // Check for a fixed point.
-                if !self.bv.current_environment.subset(&self.out_state[&bb]) {
+                if !self.out_state[&last_block].subset(&old_state[&last_block]) {
                     // There is some path for which self.bv.current_environment.value_at(path) includes
-                    // a value this is not present in self.out_state[bb].value_at(path), so any block
+                    // a value this is not present in self.out_state[last_block].value_at(path), so any block
                     // that used self.out_state[bb] as part of its input state now needs to get reanalyzed.
                     changed = true;
                 }
```

Comment:
Could the last block be an arbitrary block in the loop body? Ideally, I think we need to check every block that has an edge that escapes the loop.

Reply:
The loop body blocks are topologically sorted, so the last block is not entirely arbitrary. I think you are right to wonder what happens when there are multiple candidates for "last block" status. The block we end up choosing may well reach a fixed point before some of the other candidates. I don't think that is going to cause soundness issues, but it could decrease precision. Probably worth thinking some more and creating an issue, but I don't think I'm going to get around to deep thinking about it for quite some time to come.